capsules/src/virtual_aes_ccm.rs

```diff
@@ -71,8 +102,153 @@ enum CCMState {
     Encrypt,
 }
 
-pub struct AES128CCM<'a, A: AES128<'a> + AES128Ctr + AES128CBC> {
+// to cache up the function parameters of the crypt() function
+struct CryptFunctionParameters {
+    buf: &'static mut [u8],
+    a_off: usize,
+    m_off: usize,
+    m_len: usize,
+    mic_len: usize,
+    confidential: bool,
+    encrypting: bool,
+}
+
+impl CryptFunctionParameters {
+    pub fn new(
+        buf: &'static mut [u8],
+        a_off: usize,
+        m_off: usize,
+        m_len: usize,
+        mic_len: usize,
+        confidential: bool,
+        encrypting: bool,
+    ) -> CryptFunctionParameters {
+        CryptFunctionParameters {
+            buf: buf,
+            a_off: a_off,
+            m_off: m_off,
+            m_len: m_len,
+            mic_len: mic_len,
+            confidential: confidential,
+            encrypting: encrypting,
+        }
+    }
+}
+
+pub struct MuxAES128CCM<'a, A: AES128<'a> + AES128Ctr + AES128CBC> {
     aes: &'a A,
+    clients: List<'a, VirtualAES128CCM<'a, A>>,
+    inflight: OptionalCell<&'a VirtualAES128CCM<'a, A>>,
+    deferred_caller: &'a DynamicDeferredCall,
+    handle: OptionalCell<DeferredCallHandle>,
+}
+
+impl<'a, A: AES128<'a> + AES128Ctr + AES128CBC> MuxAES128CCM<'a, A> {
+    pub fn new(aes: &'a A, deferred_caller: &'a DynamicDeferredCall) -> MuxAES128CCM<'a, A> {
+        // aes.enable(); // enable the hardware
```

Comment:
Why is this commented out?

Reply:
Hi Alistair, 
The device should be globally enabled in the board's related file instead of here. It's same semantic for the original AES CCM driver. 
Regards, 
Yancheng Ou

Reply:
I think this line should either be deleted or commented in. I think it makes the most sense to just enable it here, even if the original driver did not do it that way. That helps prevent board authors from forgetting the enable call

Reply:
OK~ then let me make another commit and uncomment the line~ 
Sorry for this... 

Regards, 
Yancheng Ou

Reply:
It's uncommented in recent commit. And the tests are passed. 

Regards, 
Yancheng Ou